src/dune_pkg/opam_solver.ml

```diff
@@ -109,18 +113,33 @@ module Context_for_dune = struct
     | Unavailable -> Format.pp_print_string f "Availability condition not satisfied"
   ;;
 
-  let opam_version_compare =
-    let opam_file_compare_by_version =
-      let opam_package_version_compare a b =
-        OpamPackage.Version.compare a b |> Ordering.of_int
-      in
-      fun a b ->
-        opam_package_version_compare (OpamFile.OPAM.version a) (OpamFile.OPAM.version b)
+  (* Compare two packages where the "least" of the two packages is the
+     one that the solver should prefer. It is only meaningful to call
+     this function with two different versions of the same
+     package. This is sensitive to the configured version
+     preference. E.g. if the version preference is to prefer newer
+     packages then packages versions that are numerically greater will
+     be treated as less than versions that are numerically lower. This
+     comparison also accounts for the avoid-version flag by treating
+     any version with this flag set as greater than any version
+     without this flag so that the solver will prefer package versions
+     without this flag. *)
+  let opam_version_compare t a b =
+    let opam_package_version_compare a b =
+      OpamPackage.Version.compare a b |> Ordering.of_int
     in
-    fun t ->
-      match t.version_preference with
-      | Oldest -> opam_file_compare_by_version
-      | Newest -> Ordering.reverse opam_file_compare_by_version
+    match opam_file_is_avoid_version a, opam_file_is_avoid_version b with
+    | true, true | false, false ->
+      (* If both packages have the same setting of their avoid-version
+         flag then compare their versions. *)
+      let ordering =
+        opam_package_version_compare (OpamFile.OPAM.version a) (OpamFile.OPAM.version b)
+      in
+      (match t.version_preference with
+       | Oldest -> ordering
+       | Newest -> Ordering.opposite ordering)
+    | false, true -> Lt
+    | true, false -> Gt
```

Comment:
Is this the same as comparing `(opam_file_is_avoid_version, OpamPackage.Version.t)`? If it is, then `Tuple.T2.compare` would make this more readable.

Reply:
I've pushed onto the PR with this suggestion, it does make it a bit nicer.